pkg/storagev2/internal/upload/uploadstore_test.go

```diff
@@ -0,0 +1,267 @@
+		},
+	}, {
+		name: "random values",
+		test: &storagetest.ItemMarshalAndUnmarshalTest{
+			Item: &upload.PushItem{
+				Timestamp: rand.Uint64(),
+				Address:   swarmtesting.RandomAddress(),
+				TagID:     rand.Uint64(),
+			},
+			Factory: func() storage.Item { return new(upload.PushItem) },
+		},
+	}, {
+		name: "invalid size",
+		test: &storagetest.ItemMarshalAndUnmarshalTest{
+			Item: &storagetest.ItemStub{
+				MarshalBuf:   []byte{0xFF},
+				UnmarshalBuf: []byte{0xFF},
+			},
+			Factory:      func() storage.Item { return new(upload.PushItem) },
+			UnmarshalErr: upload.ErrPushItemUnmarshalInvalidSize,
+		},
+	}}
+
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			t.Parallel()
+
+			storagetest.TestItemMarshalAndUnmarshal(t, tc.test)
+		})
+	}
+}
+
+func TestChunkPutter(t *testing.T) {
+	t.Parallel()
+
+	now := func() time.Time { return time.Unix(1234567890, 0) }
+	upload.ReplaceTimeNow(now)
+	t.Cleanup(func() { upload.ReplaceTimeNow(time.Now) })
+
+	ctx, cancel := context.WithCancel(context.Background())
+	t.Cleanup(cancel)
+
+	ts := &testStorage{
+		ctx:        ctx,
+		indexStore: inmem.New(),
+		chunkStore: inmemchunkstore.New(),
+	}
+	t.Cleanup(func() {
+		if err := ts.Storage().Close(); err != nil {
+			t.Errorf("Storage().Close(): unexpected error: %v", err)
+		}
+		if err := ts.ChunkStore().Close(); err != nil {
+			t.Errorf("ChunkStore().Close(): unexpected error: %v", err)
+		}
+	})
+
+	tag := uint64(1)
+	putter, err := upload.ChunkPutter(ts, tag)
+	if err != nil {
+		t.Fatalf("upload.ChunkPutter(...): unexpected error: %v", err)
+	}
+	chunk := swarm.NewChunk(swarm.NewAddress([]byte("1234")), []byte("data"))
+
+	t.Run("put new chunk", func(t *testing.T) {
+		exists, err := putter.Put(context.TODO(), chunk)
+		if err != nil {
+			t.Fatalf("Put(...): unexpected error: %v", err)
+		}
+		if exists {
+			t.Fatal("Put(...): chunk should not exist")
+		}
+	})
+
+	t.Run("put existing chunk", func(t *testing.T) {
+		exists, err := putter.Put(context.TODO(), chunk)
+		if err != nil {
+			t.Fatalf("Put(...): unexpected error: %v", err)
+		}
+		if !exists {
+			t.Fatal("Put(...): chunk should exist")
+		}
+	})
+
+	t.Run("verify internal state", func(t *testing.T) {
+		has, err := ts.Storage().Has(&upload.TagIDAddressItem{
+			TagID:   tag,
+			Address: chunk.Address(),
+		})
+		if err != nil {
+			t.Fatalf("Has(...): unexpected error: %v", err)
+		}
+		if !has {
+			t.Fatal("Has(...): item not found")
+		}
+
+		has, err = ts.Storage().Has(&upload.PushItem{
+			Timestamp: uint64(now().Unix()),
+			Address:   chunk.Address(),
+			TagID:     tag,
+		})
+		if err != nil {
+			t.Fatalf("Has(...): unexpected error: %v", err)
+		}
+		if !has {
+			t.Fatalf("Has(...): item not found")
+		}
+
+		have, err := ts.ChunkStore().Get(context.TODO(), chunk.Address())
+		if err != nil {
+			t.Fatalf("Get(...): unexpected error: %v", err)
+		}
+		if diff := cmp.Diff(chunk, have); diff != "" {
```

Comment:
We have a chunk.Equal function which would be better here.

Reply:
Done.